Source/Utility/Diagnostics/DiagProbe.cpp

```diff
@@ -0,0 +1,328 @@
+          // Store grid size, cell corner and probe index
+          dx_finest_lev_probe[idim] = dx[idim];
+          m_probe_idx[idim] = idx_lev[idim];
+          x_low_cell[idim] =
+            std::floor(
+              (m_probe_loc[idim] - (problo[idim] + dx[idim] * 0.5)) /
+              dx[idim]) *
+              dx[idim] +
+            (problo[idim] + dx[idim] * 0.5);
+          low_cell_idx[idim] =
+            static_cast<int>((x_low_cell[idim] - (problo[idim])) / dx[idim]);
+        }
+        probe_found = true;
+      }
+    }
+  }
+
+  if (!probe_found) {
+    amrex::Abort(
+      "\nUnable to find the probe location. There seems to be something wrong");
+  }
+
+  // We are storing only the single box which contains the probe
+  m_probebox.resize(1);
+  m_probeboxDM.resize(1);
+  m_dmConvert.resize(1);
+
+  amrex::Vector<int> pmap;
+  amrex::BoxList bl(a_grids[m_finest_level_probe].ixType());
+  bl.reserve(1);
+  amrex::Vector<int> dmConvertLev;
+  auto cBox = a_grids[m_finest_level_probe][m_box_probe_num];
+  bl.push_back(cBox);
+  pmap.push_back(a_dmap[m_finest_level_probe][m_box_probe_num]);
+  dmConvertLev.push_back(m_box_probe_num);
+  m_dmConvert[0] = dmConvertLev;
+  m_probebox[0].define(bl);
+  m_probeboxDM[0].define(pmap);
+}
+
+void
+DiagProbe::processDiag(
+  int a_nstep,
+  const amrex::Real& a_time,
+  const amrex::Vector<const amrex::MultiFab*>& a_state,
+  const amrex::Vector<std::string>& a_varNames)
+{
+  // since we are only taking the single box, just create a single multifab
+  amrex::Vector<amrex::MultiFab> planeData(1);
+  planeData[0].define(
+    m_probebox[0], m_probeboxDM[0], static_cast<int>(m_fieldNames.size()), 0);
+  m_values_at_probe.resize(m_fieldIndices.size());
+  std::fill(m_values_at_probe.begin(), m_values_at_probe.end(), 0.0);
+  amrex::Gpu::copy(
+    amrex::Gpu::hostToDevice, m_values_at_probe.begin(),
+    m_values_at_probe.end(), m_values_at_probe_d.begin());
+
+  for (amrex::MFIter mfi(planeData[0], amrex::TilingIfNotGPU()); mfi.isValid();
+       ++mfi) {
+    const auto& bx = mfi.tilebox();
+    const int state_idx = m_dmConvert[0][mfi.index()];
+    auto const& state =
+      a_state[m_finest_level_probe]->const_array(state_idx, 0);
+    auto const& plane = planeData[0].array(mfi);
+    auto* idx_d_p = m_fieldIndices_d.dataPtr();
+    amrex::Real* tmp_values_d = m_values_at_probe_d.data();
+    auto const& m_probe_idx_d = m_probe_idx;
+    auto const& low_cell_idx_d = low_cell_idx;
+
+    auto& m_probe_loc_d = m_probe_loc;
+    auto& x_low_cell_d = x_low_cell;
+    auto& dx_finest_lev_probe_d = dx_finest_lev_probe;
+    auto& m_interpType_d = m_interpType;
+
+    amrex::ParallelFor(
+      m_fieldIndices_d.size(), [=] AMREX_GPU_DEVICE(int n) noexcept {
+        amrex::Array3D<amrex::Real, 0, 1, 0, 1, 0, 1> neighbour_cells{
+          0.0}; // Neighbour cell solution values
+        int stIdx = idx_d_p[n];
+
+        if (m_interpType_d == Linear) {
```

Comment:
I would pull this if out of the parallelfor and have 2 parallelfors. This is to avoid potential thread divergence on the GPU. There's very little shared code so I think it would be easy to do.